game/minecraft_nbt.ksy

```diff
@@ -0,0 +1,227 @@
+  ) by Notch), but can also be compressed with _zlib_ or uncompressed.
+
+  This spec can only handle uncompressed NBT data, so be sure to first detect
+  what type of data you are dealing with. You can use the Unix `file` command
+  to do this:
+
+  ```shell
+  file --brief --mime-type input-unknown.nbt
+  ```
+
+  If it says:
+
+    * `application/gzip`, you can decompress it by
+      * `gunzip -c input-gzip.nbt > output.nbt` or
+      * `python3 -c "import sys, gzip; sys.stdout.buffer.write(
+        gzip.decompress(sys.stdin.buffer.read()) )" < input-gzip.nbt > output.nbt`
+    * `application/zlib`, you can use
+      * `openssl zlib -d -in input.nbt.zlib -out output.nbt` (does not work on most systems)
+      * `python3 -c "import sys, zlib; sys.stdout.buffer.write(
+        zlib.decompress(sys.stdin.buffer.read()) )" < input-zlib.nbt > output.nbt`
+    * something else (especially `image/x-pcx` and `application/octet-stream`),
+      it is most likely already uncompressed.
+
+  The file `output.nbt` generated by one of the above commands can already be
+  processed with this Kaitai Struct specification.
+
+  This spec **only** implements the Java edition format. There is also
+  a [Bedrock edition](https://wiki.vg/NBT#Bedrock_edition) NBT format,
+  which uses little-endian encoding and has a few other differences, but it isn't
+  as popular as the Java edition format.
+
+  **Implementation note:** strings in `TAG_String` are incorrectly decoded with
+  standard UTF-8, while they are encoded in [**Modified UTF-8**](
+    https://docs.oracle.com/javase/8/docs/api/java/io/DataInput.html#modified-utf-8
+  ) (MUTF-8). That's because MUTF-8 is usually not supported natively by most target
+  languages, and thus one must use external libraries to achieve a fully-compliant
+  decoder. But decoding in standard UTF-8 is still better than nothing, and
+  it usually works fine.
+
+  All Unicode codepoints with incompatible representations in MUTF-8 and UTF-8 are
+  U+0000 (_NUL_), U+D800-U+DFFF (_High_ and _Low Surrogates_) and U+10000-U+10FFFF
+  (_Supplementary_ Planes; includes e.g. emoticons, pictograms).
+  A _MUTF-8_-encoded string containing these code points cannot be successfully
+  decoded as UTF-8. The behavior in this case depends on the target language -
+  usually an exception is thrown, or the bytes that are not valid UTF-8
+  are replaced or ignored.
+
+  ---
+
+  Sample files:
+
+    * <https://wiki.vg/NBT#Download>
+    * <https://github.com/twoolie/NBT/blob/f9e892e/tests/world_test/data/scoreboard.dat>
+    * <https://github.com/chmod222/cNBT/tree/3f74b69/testdata>
+    * <https://github.com/PistonDevelopers/hematite_nbt/tree/0b85f89/tests>
+doc-ref:
+  - https://wiki.vg/NBT
+  - https://web.archive.org/web/20110723210920/https://www.minecraft.net/docs/NBT.txt
+  - https://minecraft.gamepedia.com/NBT_format
+seq:
+  - size: 0
+    if: root_type == tag::end # always `false`, but it doesn't matter
```

Comment:
Thinking about this again, would it make sense to move the `valid` check to here:

```suggestion
    valid:
      expr: root_type == tag::compound
```

and remove the `valid: tag::compound` below? This makes the intent (checking the root type at exactly this point in parsing) more obvious and avoids having to use a nonsense condition like this to force evaluation of an instance.

Or if you want to keep it this way, please change the condition to

```suggestion
    if: root_type == tag::end or true # Force evaluation of root_type's valid check
```

so that it doesn't matter if the nonsense condition is true or false.

Reply:
> Thinking about this again, would it make sense to move the `valid` check to here [...] and remove the `valid: tag::compound` below?

I'm OK with that. It does the job, and yeah, you're right that it makes the weird `size: 0` member less cryptic and hacky.

Reply:
Addressed in https://github.com/kaitai-io/kaitai_struct_formats/pull/404/commits/e37fbbbd100a1279151fe1bc83606c110e32dfb5 and it really simplifies things a lot! Thanks for this good suggestion.

Reply:
Well, I'm not sure now. I have some new findings about this snippet:

```ksy
  - size: 0
    valid:
      expr: root_type == tag::compound
```

1. `valid/expr` in unnamed fields (i.e. without `id`) will apparently not work properly in any target language. For some reason, the compiler generates this kind of code everywhere:

   ```javascript
     MinecraftNbt.prototype._read = function() {
       this._unnamed0 = this._io.readBytes(0);
       var _ = this.unnamed0;
   ```

   Notice that the property is named <code><b>_</b>unnamed0</code>, but the assignment to the underscore (`_`) variable references `unnamed0` without the underscore prefix. So I guess the code won't even compile in static-typed languages, and in dynamic-typed it'll work only as long as the `_` alias is not used in the expression.

2. The spec using `valid/expr` won't even compile with KSC for Go and Lua, whereas if it's only using other `valid` subkeys (`eq`, `min`, `any-of`, ...), it will.

---

The first thing is pretty significant, the second not so much, but still perhaps we could wait with `valid/expr` until the next release when it's fixed for Go and Lua. What do you think, @dgelessus?

Reply:
oof, I should have checked this better before suggesting it, heh. The name problem at least could be fixed by giving the zero-size field an `id`, right? Arguably that would be the right thing to do anyway, because `id`-less fields are supposed to be for data with unknown meaning, which this isn't.

There's no good way to avoid the second issue though, other than just not using `valid/expr`. I suppose for now we could revert that change - not sure if that's worth it or not.

Reply:
> The name problem at least could be fixed by giving the zero-size field an `id`, right? Arguably that would be the right thing to do anyway, because `id`-less fields are supposed to be for data with unknown meaning, which this isn't.

Yeah, let's do that.